lgc/patch/MeshTaskShader.cpp

```diff
@@ -2664,35 +2707,46 @@ Value *MeshTaskShader::getMeshLocalInvocationId() {
+      m_meshLocalInvocationId = localInvocationId;
+    } else {
+      // localInvocationIndex = localInvocationId.z * workgroupSize.x * workgroupSize.y +
+      //                        localInvocationId.y * workgroupSize.x + localInvocationId.x
+      //
+      // localInvocationId.z = localInvocationIndex / workgroupSize.x * workgroupSize.y
+      // localInvocationId.y = (localInvocationIndex - workgroupSize.x * workgroupSize.y * localInvocationId.z) /
+      //                       workgroupSize.x
+      // localInvocationId.x = (localInvocationIndex - workgroupSize.x * workgroupSize.y * localInvocationId.z) -
+      //                       workgroupSize.x * localInvocationId.y
+      const auto &meshMode = m_pipelineState->getShaderModes()->getMeshShaderMode();
+      auto localInvocationIndex = getMeshLocalInvocationIndex();
+
+      auto workgroupSizeX = m_builder.getInt32(meshMode.workgroupSizeX);
+      auto workgroupSizeXMulY = m_builder.getInt32(meshMode.workgroupSizeX * meshMode.workgroupSizeY);
 
-    auto localInvocationIdZ = m_builder.CreateUDiv(localInvocationIndex, workgroupSizeXMulY);
+      auto localInvocationIdZ = m_builder.CreateUDiv(localInvocationIndex, workgroupSizeXMulY);
 
-    auto diff = m_builder.CreateMul(workgroupSizeXMulY, localInvocationIdZ);
-    diff = m_builder.CreateSub(localInvocationIndex, diff);
-    auto localInvocationIdY = m_builder.CreateUDiv(diff, workgroupSizeX);
+      auto diff = m_builder.CreateMul(workgroupSizeXMulY, localInvocationIdZ);
+      diff = m_builder.CreateSub(localInvocationIndex, diff);
+      auto localInvocationIdY = m_builder.CreateUDiv(diff, workgroupSizeX);
 
-    auto localInvocationIdX = m_builder.CreateMul(workgroupSizeX, localInvocationIdY);
-    localInvocationIdX = m_builder.CreateSub(diff, localInvocationIdX);
+      auto localInvocationIdX = m_builder.CreateMul(workgroupSizeX, localInvocationIdY);
+      localInvocationIdX = m_builder.CreateSub(diff, localInvocationIdX);
 
-    Value *localInvocationId = UndefValue::get(FixedVectorType::get(m_builder.getInt32Ty(), 3));
-    localInvocationId = m_builder.CreateInsertElement(localInvocationId, localInvocationIdX, static_cast<uint64_t>(0));
-    localInvocationId = m_builder.CreateInsertElement(localInvocationId, localInvocationIdY, 1);
-    localInvocationId = m_builder.CreateInsertElement(localInvocationId, localInvocationIdZ, 2);
+      Value *localInvocationId = UndefValue::get(FixedVectorType::get(m_builder.getInt32Ty(), 3));
+      localInvocationId =
+          m_builder.CreateInsertElement(localInvocationId, localInvocationIdX, static_cast<uint64_t>(0));
+      localInvocationId = m_builder.CreateInsertElement(localInvocationId, localInvocationIdY, 1);
+      localInvocationId = m_builder.CreateInsertElement(localInvocationId, localInvocationIdZ, 2);
 
-    m_meshLocalInvocationId = localInvocationId;
+      m_meshLocalInvocationId = localInvocationId;
```

Comment:
How about moving most of this code into `initWaveTheadInfo` so that we can rely on `m_waveThreadInfo.localInvocationId{X,Y,Z}` always being initialized?

At some point we should also consider unifying what happens here with the CS local invocation ID swizzling that's implemented in PatchInOutImportExport. This is currently difficult to do though.

Reply:
> so that we can rely on m_waveThreadInfo.localInvocationId{X,Y,Z} always being initialized?

OK. I will apply similar changes to workgroupId as well.

> we should also consider unifying what happens here with the CS local invocation ID swizzling that's implemented in PatchInOutImportExport

I suppose we need a new call as a getter of local invocation ID in the future and lower it in later passes with the replacement of "getting current local invocation ID"